src/binary-reader-objdump.cc

```diff
@@ -956,34 +964,32 @@ Result BinaryReaderObjdumpDisassemble::OnOpcodeBlockSig(Type sig_type) {
   return Result::Ok;
 }
 
-enum class InitExprType {
-  I32,
-  F32,
-  I64,
-  F64,
-  V128,
-  Global,
-  FuncRef,
-  // TODO: There isn't a nullref anymore, this just represents ref.null of some
-  // type T.
-  NullRef,
+struct I32Value {
+  uint32_t value;
+};
+
+struct I64Value {
+  uint64_t value;
+};
+
+struct F32Value {
+  uint32_t value;
+};
+
+struct F64Value {
+  uint64_t value;
+};
+
+struct GlobalIndex {
+  Index value;
 };
 
 struct InitInst {
   Opcode opcode;
-  union {
-    Index index;
-    uint32_t i32;
-    uint32_t f32;
-    uint64_t i64;
-    uint64_t f64;
-    v128 v128_v;
-    Type type;
-  } imm;
+  std::variant<I32Value, I64Value, F32Value, F64Value, v128, GlobalIndex> data;
```

Comment:
Why not continue to call this `imm` which seems more descriptive?

Also, maybe we can call all the possible variantes accordingly: ImmI32, ImmI64, ImmF64, ImmV128, ImmIndex.

Reply:
Also can we avoid all the `.value` accessors below if we use typedefs here instead of nested structs:

```
typedef ImmI32 uint32_t;
```

?

Reply:
Thanks for the comments Sam!
I've returned `imm` name and renamed structs.

Actually I can't reduce `.value` because in that case variant will have equal types. For example, we represent `i64` and `f64` as `uint64_t`. In that case we can't use `std::visit` and `std::get<Type>` because variant's options aren't unique, so one possible solution is to use indices:

```
using ImmI32 = uint32_t;
using ImmI64 = uint64_t;
using ImmF32 = uint32_t;
using ImmF64 = uint64_t;
using ImmV128 = v128;
using ImmIndex = Index;

constexpr std::size_t ImmI32TypeIdx = 0;
constexpr std::size_t ImmI64TypeIdx = 1;
constexpr std::size_t ImmF32TypeIdx = 2;
constexpr std::size_t ImmF64TypeIdx = 3;
constexpr std::size_t ImmV128TypeIdx = 4;
constexpr std::size_t ImmIndexTypeIdx = 5;

```

For that case we should manually keep them in sync with the order of types in the variant.
Moreover, beautiful `std::visit` for this case can't be used and then should be replaced with a chain of if-then-else, so I think structs are ok, wdyt?

Reply:
I don't understand the "variant's options aren't unique" part?  Aren't different typedefs considered unique types?

Reply:
I mean when we have typedefs or usings like this:

```
using ImmI32 = uint32_t;
using ImmI64 = uint64_t;
using ImmF32 = uint32_t;
using ImmF64 = uint64_t;
using ImmV128 = v128;
using ImmIndex = Index;
```

then `std::variant<ImmI32, ImmI64, ImmF32, ImmF64, ImmV128, ImmIndex>` translates into 
`std::variant<uint32_t, uint64_t, uint32_t, uint64_t, v128, Index>`.
So, we can't use `std::get<ImmI32>` because it translates into `std::get<uint32_t>` because using/typedef doesn't introduce new types, it is just a type alias. In the end, `std::get<uint32_t>` is ambiguous because it can refer to a place of ImmI32 or refer to a place of ImmF32. 

The documentations for [std::get](https://en.cppreference.com/w/cpp/utility/variant/get) says: `The call is ill-formed if T is not a unique element of Types....`

Reply:
In that case maybe we can reduce the variant to just `ImmU32`, `ImmU64` and `ImmV128`?

Reply:
Actually we can, but in that case `PrintInitExpr(const InitExpr& expr)` can't distinguish between F64 and I64 without any additional enum opcode support (we removed it in this PR).

The main argument here is that we replace the (union + enum opcode) scheme with a single `std::variant` because it is more convenient to add new types and compiler can check that some options in `std::visit` are uncovered.

I don't see any memory or speed degradation from this change, at least in theory, because `std::variant` will have the same size as the previous union based solution with `enum class InitExprType`. With variant we just don't need to manually maintain InitExprType.